src/librustc/session/mod.rs

```diff
@@ -331,6 +334,105 @@ impl Session {
             &self.opts.search_paths,
             kind)
     }
+
+    pub fn init_incr_comp_session(&self,
+                                  session_dir: PathBuf,
+                                  lock_file: flock::Lock) {
+        let mut incr_comp_session = self.incr_comp_session.borrow_mut();
+
+        match *incr_comp_session {
+            IncrCompSession::Active { .. } => {
+                bug!("Trying initialize already active IncrCompSession.")
+            }
+            IncrCompSession::Finalized { .. } => {
+                bug!("Trying initialize already finalized IncrCompSession.")
+            }
+            IncrCompSession::InvalidBecauseOfErrors => {
+                bug!("Trying to initialize invalidated IncrCompSession.");
+            }
+            IncrCompSession::NotInitialized => {
+                // This is the correct state, proceed ...
+            }
+        }
+
+        *incr_comp_session = IncrCompSession::Active {
+            session_directory: session_dir,
+            lock_file: lock_file,
+        };
+    }
+
+    pub fn finalize_incr_comp_session(&self, new_directory_path: PathBuf) {
+        let mut incr_comp_session = self.incr_comp_session.borrow_mut();
+
+        match *incr_comp_session {
+            IncrCompSession::NotInitialized => {
```

Comment:
As long as I can see the current state of the session in the error message, I'm fine with that.